components/permissions/google_sign_in/google_sign_in_permission_throttle.cc

```diff
@@ -0,0 +1,149 @@
+
+using blink::URLLoaderThrottle;
+
+namespace permissions {
+
+GoogleSignInPermissionThrottle::GoogleSignInPermissionThrottle(
+    const content::WebContents::Getter wc_getter,
+    const GURL initial_url,
+    scoped_refptr<HostContentSettingsMap> settings_map)
+    : wc_getter_(wc_getter),
+      initial_url_(initial_url),
+      settings_map_(settings_map) {}
+
+void OnPermissionRequestStatus(
+    content::NavigationEntry* pending_entry,
+    content::WebContents* contents,
+    const GURL& request_initiator_url,
+    scoped_refptr<HostContentSettingsMap> content_settings,
+    URLLoaderThrottle::Delegate* delegate,
+    const std::vector<blink::mojom::PermissionStatus>& permission_statuses) {
+  HandleBraveGoogleSignInPermissionStatus(
+      contents->GetBrowserContext(), request_initiator_url, content_settings,
+      permission_statuses);
+  DCHECK_EQ(1u, permission_statuses.size());
+  const auto status = permission_statuses[0];
+  // Check if current pending navigation is the one we started out with.
+  // This is done to prevent us from accessing a deleted Delegate, if
+  // the user navigated away while the prompt was still up, or closed the
+  // window
+  if (pending_entry != contents->GetController().GetPendingEntry()) {
+    return;
+  }
+  if (status == blink::mojom::PermissionStatus::GRANTED) {
+    delegate->Resume();
+  } else if (status == blink::mojom::PermissionStatus::DENIED) {
+    delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+  }
+  // In case of ASK, we need to be careful because delegate may be deleted
+  return;
+}
+
+std::unique_ptr<blink::URLLoaderThrottle>
+GoogleSignInPermissionThrottle::MaybeCreateThrottleFor(
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter wc_getter,
+    HostContentSettingsMap* content_settings) {
+  if (!IsGoogleSignInFeatureEnabled()) {
+    return nullptr;
+  }
+
+  if (request.resource_type !=
+      static_cast<int>(blink::mojom::ResourceType::kMainFrame)) {
+    return nullptr;
+  }
+
+  const auto& request_url = request.url;
+
+  // We might not have a request_initiator_url if the request is coming from
+  // a top-level navigation and it's a redirect. We still want to create the
+  // throttle.
+  if (!request_url.SchemeIsHTTPOrHTTPS()) {
+    return nullptr;
+  }
+
+  return std::make_unique<GoogleSignInPermissionThrottle>(
+      wc_getter, request_url,
+      base::WrapRefCounted<HostContentSettingsMap>(content_settings));
+}
+
+GoogleSignInPermissionThrottle::~GoogleSignInPermissionThrottle() = default;
+
+void GoogleSignInPermissionThrottle::DetachFromCurrentSequence() {}
+
+void OnPermissionDeny(URLLoaderThrottle::Delegate* delegate) {
+  delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+  return;
+}
+
+void HandleRequest(bool* defer,
+                   const GURL& request_url,
+                   const GURL& request_initiator_url,
+                   const content::WebContents::Getter contents_getter,
```

Comment:
`const content::WebContents::Getter& contents_getter`

it's ok to pass references as parameters. 
it's not ok to store them as class members.